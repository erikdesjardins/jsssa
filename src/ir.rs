use swc_atoms::JsWord;

pub use self::print::print;
pub use self::ref_::{Mutable, Ref, RefType, SSA};

mod print;
mod ref_;
pub mod scope;
pub mod visit;

#[derive(Debug)]
pub struct Block(pub Vec<Stmt>);

#[derive(Debug)]
pub enum Stmt {
    Expr {
        target: Ref<SSA>,
        expr: Expr,
    },
    DeclareMutable {
        target: Ref<Mutable>,
        val: Ref<SSA>,
    },
    WriteMutable {
        target: Ref<Mutable>,
        val: Ref<SSA>,
    },
    WriteGlobal {
        target: JsWord,
        val: Ref<SSA>,
    },
    WriteMember {
        obj: Ref<SSA>,
        prop: Ref<SSA>,
        val: Ref<SSA>,
    },
    Return {
        val: Ref<SSA>,
    },
    Throw {
        val: Ref<SSA>,
    },
    Break,
    Continue,
    Debugger,
    Block {
        body: Block,
    },
    Loop {
        body: Block,
    },
    ForEach {
        kind: ForKind,
        init: Ref<SSA>,
        body: Block,
    },
    IfElse {
        cond: Ref<SSA>,
        cons: Block,
        alt: Block,
    },
    Try {
        body: Block,
        catch: Block,
        finally: Box<Block>,
    },
}

#[derive(Debug)]
pub enum Expr {
    Bool {
        value: bool,
    },
    Number {
        value: f64,
    },
    String {
        value: JsWord,
        has_escape: bool,
    },
    Null,
    Undefined,
    This,
    Read {
        source: Ref<SSA>,
    },
    ReadMutable {
        source: Ref<Mutable>,
    },
    ReadGlobal {
        source: JsWord,
    },
    ReadMember {
        obj: Ref<SSA>,
        prop: Ref<SSA>,
    },
    Array {
        elems: Vec<Option<(EleKind, Ref<SSA>)>>,
    },
    Object {
        props: Vec<(PropKind, Ref<SSA>, Ref<SSA>)>,
    },
    RegExp {
        regex: JsWord,
        has_escape: bool,
        flags: JsWord,
    },
    Unary {
        op: UnaryOp,
        val: Ref<SSA>,
    },
    Binary {
        op: BinaryOp,
        left: Ref<SSA>,
        right: Ref<SSA>,
    },
    Delete {
        obj: Ref<SSA>,
        prop: Ref<SSA>,
    },
    Yield {
        kind: YieldKind,
        val: Ref<SSA>,
    },
    Await {
        val: Ref<SSA>,
    },
    Call {
        kind: CallKind,
        func: Ref<SSA>,
        args: Vec<(EleKind, Ref<SSA>)>,
    },
    Function {
        kind: FnKind,
        body: Block,
    },
    CurrentFunction,
    Argument {
        index: usize,
    },
}

#[derive(Debug)]
pub enum UnaryOp {
    Plus,
    Minus,
    Not,
    Tilde,
    Typeof,
    Void,
}

#[derive(Debug)]
pub enum BinaryOp {
    EqEq,
    NotEq,
    StrictEq,
    NotStrictEq,
    Lt,
    LtEq,
    Gt,
    GtEq,
    ShiftLeft,
    ShiftRight,
    ShiftRightZero,
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    BitOr,
    BitXor,
    BitAnd,
    Exp,
    In,
    Instanceof,
}

#[derive(Debug)]
pub enum ForKind {
    In,
    Of,
}

#[derive(Debug)]
pub enum EleKind {
    Single,
    Spread,
}

#[derive(Debug)]
pub enum PropKind {
    Simple,
    Get,
    Set,
}

#[derive(Debug)]
pub enum CallKind {
    Call,
    New,
}

#[derive(Debug)]
pub enum FnKind {
    Func { is_async: bool, is_generator: bool },
    Arrow { is_async: bool },
}

#[derive(Debug)]
pub enum YieldKind {
    Single,
    Delegate,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn stmt_is_512_bits() {
        assert_eq!(std::mem::size_of::<Stmt>(), 64);
    }
}
