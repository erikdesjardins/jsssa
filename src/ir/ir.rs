use std::fmt::{self, Debug};
use std::marker::PhantomData;
use std::rc::Rc;
use std::sync::atomic::{AtomicUsize, Ordering};

// references can either be SSA or mutable (needed to model closures)
pub trait RefType {}
impl RefType for SSA {}
impl RefType for Mutable {}
#[derive(Clone, Debug)]
pub enum SSA {}
#[derive(Clone, Debug)]
pub enum Mutable {}

#[derive(Clone)]
pub enum Ref<T: RefType> {
    Dead,
    Live(LiveRef<T>),
}

impl<T: RefType> Debug for Ref<T> {
    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        let mut f = f.debug_tuple("Ref");
        match self {
            Ref::Dead => {
                f.field(&"Dead");
            }
            Ref::Live(LiveRef(inner)) => {
                f.field(&inner.id);
                f.field(&inner.name_hint);
            }
        }
        f.finish()
    }
}

impl<T: RefType> Ref<T> {
    pub fn new(name_hint: String) -> Self {
        Ref::Live(LiveRef::new(name_hint))
    }

    pub fn dead() -> Self {
        Ref::Dead
    }
}

#[derive(Clone)]
struct LiveRef<T: RefType>(Rc<LiveRefInner<T>>);

struct LiveRefInner<T: RefType> {
    id: usize,
    name_hint: String,
    _t: PhantomData<T>,
}

impl<T: RefType> PartialEq for LiveRef<T> {
    fn eq(&self, other: &Self) -> bool {
        // compare only by id, which is unique by construction
        self.0.id == self.0.id
    }
}

impl<T: RefType> LiveRef<T> {
    fn new(name_hint: String) -> Self {
        static ID_COUNTER: AtomicUsize = AtomicUsize::new(1);

        LiveRef(Rc::new(LiveRefInner {
            id: ID_COUNTER.fetch_add(1, Ordering::Relaxed),
            name_hint,
            _t: PhantomData,
        }))
    }

    pub fn name_hint(&self) -> &str {
        &self.0.name_hint
    }

    pub fn might_be_used(&self) -> bool {
        Rc::strong_count(&self.0) > 1
    }
}

#[derive(Debug)]
pub struct Block {
    pub directives: Vec<String>,
    pub children: Vec<Stmt>,
}

impl Block {
    pub fn empty() -> Self {
        Block {
            directives: vec![],
            children: vec![],
        }
    }
}

#[derive(Debug)]
pub enum Stmt {
    Expr(Expr),
    Assign(Ref<SSA>, Expr),
    AssignBinding(Ref<Mutable>, Ref<SSA>),
    AssignGlobal(String, Ref<SSA>),
    Return(Ref<SSA>),
    Throw(Ref<SSA>),
    Debugger,
    Block(Box<Block>),
    IfElse(Ref<SSA>, Box<Block>, Box<Block>),
    Try(
        Box<Block>,
        Option<(Ref<Mutable>, Box<Block>)>,
        Option<Box<Block>>,
    ),
}

#[derive(Debug)]
pub enum Expr {
    Bool(bool),
    Number(f64),
    String(String),
    Null,
    Undefined,
    This,
    Read(Ref<SSA>),
    ReadBinding(Ref<Mutable>),
    ReadGlobal(String),
    Array(Vec<Option<(EleKind, Ref<SSA>)>>),
    Object(Vec<(PropKind, Ref<SSA>, Ref<SSA>)>),
    RegExp(String, String),
    Unary(UnaryOp, Ref<SSA>),
    Binary(BinaryOp, Ref<SSA>, Ref<SSA>),
    Call(CallKind, Ref<SSA>, Vec<Ref<SSA>>),
    Function(FnKind, Option<String>, Vec<Ref<Mutable>>, Box<Block>),
    Yield(YieldKind, Ref<SSA>),
    Await(Ref<SSA>),
}

#[derive(Debug)]
pub enum UnaryOp {
    Plus,
    Minus,
    Not,
    Tilde,
    Typeof,
    Void,
    Delete,
}

#[derive(Debug)]
pub enum BinaryOp {
    Eq,
    NotEq,
    StrictEq,
    NotStrictEq,
    LT,
    LTE,
    GT,
    GTE,
    ShiftLeft,
    ShiftRight,
    ShiftRightZero,
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    BitOr,
    BitXor,
    BitAnd,
    In,
    Instanceof,
}

#[derive(Debug)]
pub enum EleKind {
    Single,
    Spread,
}

#[derive(Debug)]
pub enum PropKind {
    Simple,
    Get,
    Set,
}

#[derive(Debug)]
pub enum CallKind {
    Call,
    New,
}

#[derive(Debug)]
pub enum FnKind {
    Func { async: bool, gen: bool },
    Arrow { async: bool },
}

#[derive(Debug)]
pub enum YieldKind {
    Single,
    Delegate,
}
