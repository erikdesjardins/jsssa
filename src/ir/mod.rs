mod types;
mod ref_;

pub use self::types::*;
pub use self::ref_::*;
