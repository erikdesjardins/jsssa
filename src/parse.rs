use std::env;
use std::fs::File;
use std::process::{Command, Stdio};
use std::io::Write;

use failure::{Error, ResultExt};
use rand::{self, Rng};
use serde_json;

use ast;

static BABYLON_BIN: &[u8] = include_bytes!("../vendor/babylon.js");

pub fn parse(js: &str) -> Result<ast::File, Error> {
    let mut path = env::temp_dir();
    let random_name: String = rand::thread_rng().gen_ascii_chars().take(32).collect();
    path.push(random_name);

    File::create(&path)?
        .write_all(BABYLON_BIN)?;

    let mut child = Command::new("node")
        .arg(&path)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()?;

    write!(child.stdin.as_mut().unwrap(), "{}", js)?;

    let out = child.wait_with_output()?;

    if !out.status.success() {
        let stderr = String::from_utf8(out.stderr)?;
        // skip error throwing header
        let stderr = stderr
            .lines()
            .skip(3)
            .collect::<Vec<_>>()
            .join("\n");

        return Err(ParseError { stderr })?;
    }

    let stdout = String::from_utf8(out.stdout)?;
    let ast = serde_json::from_str(&stdout).context("failed to parse babylon output")?;

    Ok(ast)
}

#[derive(Fail, Debug)]
#[fail(display = "JS parse error: {}", stderr)]
struct ParseError {
    stderr: String
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn basic_foo() {
        use ast::*;

        assert_eq!(
            parse(
                r#"
                    'use strict';
                    function foo(x) {
                        return x + 1;
                    }
                "#
            ).unwrap(),
            File::new(Program::new(
                vec![
                    FunctionDeclaration::new(
                        Identifier::new("foo".to_owned()),
                        vec![Identifier::new("x".to_owned()).into()],
                        BlockStatement::new(
                            vec![
                                ReturnStatement::new(Some(
                                    BinaryExpression::new(
                                        BinaryOperator::Add,
                                        Box::new(Identifier::new("x".to_owned()).into()),
                                        Box::new(NumericLiteral::new(1.0).into()),
                                    ).into(),
                                )).into(),
                            ],
                            vec![],
                        ),
                        false,
                        false,
                    ).into(),
                ],
                vec![
                    Directive::new(DirectiveLiteral::new("use strict".to_owned())),
                ],
                SourceType::Script,
            ))
        )
    }
}
