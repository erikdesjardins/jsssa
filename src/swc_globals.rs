/// A token proving that SWC's globals are initialized.
pub struct Initialized(());

/// Initializes the globals used by SWC and provides a token as proof.
///
/// A higher-rank lifetime ensures the token cannot escape the closure.
pub fn with<R>(f: impl FnOnce(&'_ Initialized) -> R) -> R {
    swc_common::GLOBALS.set(&swc_common::Globals::new(), || f(&Initialized(())))
}
