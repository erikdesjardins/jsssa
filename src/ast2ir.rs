use std::collections::HashMap;

use ast;
use ir;
use util::*;

type ScopeMap = Map<String, ir::Ref<ir::Mutable>>;

pub fn convert(ast: ast::File) -> ir::Block {
    let ast::File {
        program:
            ast::Program {
                body,
                directives,
                source_type: _,
            },
    } = ast;

    convert_block(body, directives, &HashMap::default())
}

fn convert_block(
    body: Vec<ast::Statement>,
    directives: Vec<ast::Directive>,
    parent_scopes: &ScopeMap,
) -> ir::Block {
    let mut scope = parent_scopes.clone();

    let children = body.into_iter()
        .flat_map(|stmt| convert_statement(stmt, &mut scope))
        .collect();

    ir::Block {
        directives: directives.into_iter().map(|d| d.value.value).collect(),
        children,
    }
}

fn convert_statement(stmt: ast::Statement, scope: &mut ScopeMap) -> Vec<ir::Stmt> {
    use ast::Statement::*;

    match stmt {
        ExpressionStatement(ast::ExpressionStatement { expression }) => {
            let (mut exprs, last_expr) = convert_expression(expression, scope);
            exprs.push(ir::Stmt::Expr(last_expr));
            exprs
        },
        BlockStatement(ast::BlockStatement { body, directives }) =>
            vec![ir::Stmt::Block(box convert_block(body, directives, scope))],
        EmptyStatement(ast::EmptyStatement {}) =>
            vec![],
        DebuggerStatement(ast::DebuggerStatement {}) =>
            vec![ir::Stmt::Debugger],
        WithStatement(_) =>
            unimplemented!("with() statement not supported"),
        ReturnStatement(ast::ReturnStatement { argument }) => {
            let ref_ = ir::Ref::new("return_".to_string());
            match argument {
                Some(argument) => {
                    let (mut exprs, return_value) = convert_expression(argument, scope);
                    exprs.push(ir::Stmt::Assign(ref_.clone(), return_value));
                    exprs.push(ir::Stmt::Return(ref_));
                    exprs
                },
                None => {
                    vec![
                        ir::Stmt::Assign(ref_.clone(), ir::Expr::Undefined),
                        ir::Stmt::Return(ref_),
                    ]
                }
            }
        },
        LabeledStatement(_) |
        BreakStatement(_) |
        ContinueStatement(_) =>
            unimplemented!("labels not yet supported"),
        IfStatement(ast::IfStatement { test, consequent, alternate }) => {
            let ref_ = ir::Ref::new("if_".to_string());
            let (mut exprs, test_value) = convert_expression(test, scope);
            exprs.push(ir::Stmt::Assign(ref_.clone(), test_value));
            exprs.push(ir::Stmt::IfElse(
                ref_,
                {
                    let children = convert_statement(*consequent, scope);
                    box ir::Block { directives: vec![], children }
                },
                match alternate {
                    Some(alternate) => {
                        let children = convert_statement(*alternate, scope);
                        box ir::Block { directives: vec![], children }
                    },
                    None => box ir::Block::empty()
                },
            ));
            exprs
        },
        SwitchStatement(_) =>
            // remember, switch cases aren't evaluated (!) until they're checked for equality
            unimplemented!("switch statements not yet supported"),
        ThrowStatement(ast::ThrowStatement { argument }) => {
            let ref_ = ir::Ref::new("throw_".to_string());
            let (mut exprs, throw_value) = convert_expression(argument, scope);
            exprs.push(ir::Stmt::Assign(ref_.clone(), throw_value));
            exprs.push(ir::Stmt::Throw(ref_));
            exprs
        },
        TryStatement(ast::TryStatement { block, handler, finalizer }) => {
            let try = ir::Stmt::Try(
                {
                    let ast::BlockStatement { body, directives } = block;
                    box convert_block(body, directives, scope)
                },
                handler.map(|ast::CatchClause { param, body }| {
                    let ast::BlockStatement { body, directives } = body;
                    let mut catch_scope = scope.clone();
                    let ref_ = convert_pattern(param, &mut catch_scope);
                    (ref_, box convert_block(body, directives, &catch_scope))
                }),
                finalizer.map(|finalizer| {
                    let ast::BlockStatement { body, directives } = finalizer;
                    box convert_block(body, directives, scope)
                }),
            );
            vec![try]
        },
        _ => unimplemented!(),
    }
}

fn convert_expression(expr: ast::Expression, scope: &ScopeMap) -> (Vec<ir::Stmt>, ir::Expr) {
    use ast::Expression::*;

    match expr {
        Identifier(ast::Identifier { name }) => {
            let expr = match scope.get(&name) {
                Some(ref_) => ir::Expr::ReadBinding(ref_.clone()),
                None => ir::Expr::ReadGlobal(name),
            };
            (vec![], expr)
        }
        RegExpLiteral(ast::RegExpLiteral { pattern, flags }) => {
            (vec![], ir::Expr::RegExp(pattern, flags))
        }
        NullLiteral(ast::NullLiteral {}) => (vec![], ir::Expr::Null),
        StringLiteral(ast::StringLiteral { value }) => (vec![], ir::Expr::String(value)),
        BooleanLiteral(ast::BooleanLiteral { value }) => (vec![], ir::Expr::Bool(value)),
        NumericLiteral(ast::NumericLiteral { value }) => (vec![], ir::Expr::Number(value)),
        ThisExpression(ast::ThisExpression {}) => (vec![], ir::Expr::This),
        ArrowFunctionExpression(ast::ArrowFunctionExpression {
            params,
            body,
            async,
        }) => {
            use ast::BlockStmtOrExpr::*;

            let mut fn_scope = scope.clone();
            let refs = params
                .into_iter()
                .map(|param| convert_pattern(param, &mut fn_scope))
                .collect();
            let body = match *body {
                BlockStatement(block) => {
                    let ast::BlockStatement { body, directives } = block;
                    convert_block(body, directives, &fn_scope)
                }
                Expression(expr) => {
                    let ref_ = ir::Ref::new("arrow_".to_string());
                    let (mut exprs, return_value) = convert_expression(expr, &fn_scope);
                    exprs.push(ir::Stmt::Assign(ref_.clone(), return_value));
                    exprs.push(ir::Stmt::Return(ref_));
                    ir::Block {
                        directives: vec![],
                        children: exprs,
                    }
                }
            };
            let func = ir::Expr::Function(ir::FnKind::Arrow { async }, None, refs, box body);
            (vec![], func)
        }
        YieldExpression(ast::YieldExpression { argument, delegate }) => {
            let ref_ = ir::Ref::new("yield_".to_string());
            let (mut exprs, yield_value) = match argument {
                Some(argument) => convert_expression(*argument, scope),
                None => (vec![], ir::Expr::Undefined),
            };
            exprs.push(ir::Stmt::Assign(ref_.clone(), yield_value));
            let kind = if delegate {
                ir::YieldKind::Delegate
            } else {
                ir::YieldKind::Single
            };
            (exprs, ir::Expr::Yield(kind, ref_))
        }
        AwaitExpression(ast::AwaitExpression { argument }) => {
            let ref_ = ir::Ref::new("await_".to_string());
            let (mut exprs, await_value) = convert_expression(*argument, scope);
            exprs.push(ir::Stmt::Assign(ref_.clone(), await_value));
            (exprs, ir::Expr::Await(ref_))
        }
        _ => unimplemented!(),
    }
}

fn convert_pattern(pat: ast::Pattern, scope: &mut ScopeMap) -> ir::Ref<ir::Mutable> {
    use ast::Pattern::*;

    match pat {
        Identifier(ast::Identifier { name }) => {
            let ref_ = ir::Ref::new(name.clone());
            scope.insert(name, ref_.clone());
            ref_
        }
        MemberExpression(_) | ObjectPattern(_) | ArrayPattern(_) | RestElement(_)
        | AssignmentPattern(_) => unimplemented!("complex patterns not yet supported"),
    }
}
