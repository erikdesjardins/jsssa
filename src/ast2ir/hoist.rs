use swc_atoms::JsWord;
use swc_ecma_ast as ast;

use crate::ir;
use crate::ir::scope;

pub enum Hoist {
    Everything,
    OnlyLetConst,
}

#[inline(never)] // for better profiling
pub fn hoist_fn_decls(block: &mut [ast::Stmt]) {
    // manually hoist function declarations at the toplevel/fn scopes
    block.sort_by_key(|stmt| match stmt {
        ast::Stmt::Decl(ast::Decl::Fn(_)) => 0,
        ast::Stmt::Labeled(ast::LabeledStmt { body, .. }) => match body.as_ref() {
            ast::Stmt::Decl(ast::Decl::Fn(_)) => 0,
            _ => 1,
        },
        _ => 1,
    });
}

#[inline(never)] // for better profiling
pub fn hoist_vars(block: &[ast::Stmt], scope: &mut scope::Ast, hoist: Hoist) -> Vec<ir::Stmt> {
    hoist_block(block, scope, &hoist, Level::First)
}

enum Level {
    First,
    Below,
}

fn hoist_block(
    body: &[ast::Stmt],
    scope: &mut scope::Ast,
    hoist: &Hoist,
    level: Level,
) -> Vec<ir::Stmt> {
    match (hoist, &level) {
        (Hoist::Everything, _) | (Hoist::OnlyLetConst, Level::First) => body
            .iter()
            .flat_map(|stmt| hoist_statement(stmt, scope, hoist, &level))
            .collect(),
        (Hoist::OnlyLetConst, Level::Below) => vec![],
    }
}

fn hoist_statement(
    stmt: &ast::Stmt,
    scope: &mut scope::Ast,
    hoist: &Hoist,
    level: &Level,
) -> Vec<ir::Stmt> {
    match stmt {
        ast::Stmt::Block(ast::BlockStmt { stmts, span: _ }) => {
            hoist_block(stmts, scope, hoist, Level::Below)
        }
        ast::Stmt::With(ast::WithStmt {
            obj: _,
            body,
            span: _,
        }) => hoist_statement(body, scope, hoist, &Level::Below),
        ast::Stmt::Labeled(ast::LabeledStmt {
            label: _,
            body,
            span: _,
        }) => hoist_statement(body, scope, hoist, level),
        ast::Stmt::If(ast::IfStmt {
            test: _,
            cons,
            alt,
            span: _,
        }) => {
            let mut decls = hoist_statement(cons, scope, hoist, &Level::Below);
            if let Some(alt) = alt {
                decls.extend(hoist_statement(alt, scope, hoist, &Level::Below));
            }
            decls
        }
        ast::Stmt::Switch(ast::SwitchStmt {
            discriminant: _,
            cases,
            span: _,
        }) => cases
            .iter()
            .flat_map(
                |ast::SwitchCase {
                     test: _,
                     cons,
                     span: _,
                 }| {
                    cons.iter()
                        .flat_map(|stmt| hoist_statement(stmt, scope, hoist, &Level::Below))
                        .collect::<Vec<_>>()
                },
            )
            .collect(),
        ast::Stmt::Try(ast::TryStmt {
            block: ast::BlockStmt { stmts, span: _ },
            handler,
            finalizer,
            span: _,
        }) => {
            let mut decls = hoist_block(stmts, scope, hoist, Level::Below);
            if let Some(ast::CatchClause {
                param: _,
                body: ast::BlockStmt { stmts, span: _ },
                span: _,
            }) = handler
            {
                decls.extend(hoist_block(stmts, scope, hoist, Level::Below));
            }
            if let Some(ast::BlockStmt { stmts, span: _ }) = finalizer {
                decls.extend(hoist_block(stmts, scope, hoist, Level::Below));
            }
            decls
        }
        ast::Stmt::While(ast::WhileStmt {
            test: _,
            body,
            span: _,
        })
        | ast::Stmt::DoWhile(ast::DoWhileStmt {
            test: _,
            body,
            span: _,
        }) => hoist_statement(body, scope, hoist, &Level::Below),
        ast::Stmt::For(ast::ForStmt {
            init,
            test: _,
            update: _,
            body,
            span: _,
        }) => {
            let mut decls = match init {
                Some(ast::VarDeclOrExpr::VarDecl(var_decl)) => {
                    hoist_variable_declaration(var_decl, scope, hoist, &Level::Below)
                }
                Some(ast::VarDeclOrExpr::Expr(_)) | None => vec![],
            };
            decls.extend(hoist_statement(body, scope, hoist, &Level::Below));
            decls
        }
        ast::Stmt::ForIn(ast::ForInStmt {
            left,
            right: _,
            body,
            span: _,
        })
        | ast::Stmt::ForOf(ast::ForOfStmt {
            left,
            right: _,
            body,
            await_token: _,
            span: _,
        }) => {
            let mut decls = match left {
                ast::VarDeclOrPat::VarDecl(var_decl) => {
                    hoist_variable_declaration(var_decl, scope, hoist, &Level::Below)
                }
                // bare patterns can't introduce variables
                ast::VarDeclOrPat::Pat(_) => vec![],
            };
            decls.extend(hoist_statement(body, scope, hoist, &Level::Below));
            decls
        }
        ast::Stmt::Decl(decl) => match decl {
            ast::Decl::Fn(fn_decl) => hoist_function_declaration(fn_decl, scope, hoist),
            ast::Decl::Var(var_decl) => hoist_variable_declaration(var_decl, scope, hoist, level),
            ast::Decl::Class(_) => unimplemented!("classes not supported"),
            ast::Decl::TsInterface(_)
            | ast::Decl::TsTypeAlias(_)
            | ast::Decl::TsEnum(_)
            | ast::Decl::TsModule(_) => unreachable!(),
        },
        ast::Stmt::Expr(_)
        | ast::Stmt::Empty(_)
        | ast::Stmt::Debugger(_)
        | ast::Stmt::Return(_)
        | ast::Stmt::Break(_)
        | ast::Stmt::Continue(_)
        | ast::Stmt::Throw(_) => vec![],
    }
}

fn hoist_function_declaration(
    fn_decl: &ast::FnDecl,
    scope: &mut scope::Ast,
    hoist: &Hoist,
) -> Vec<ir::Stmt> {
    let ast::FnDecl {
        ident:
            ast::Ident {
                sym,
                span: _,
                type_ann: _,
                optional: _,
            },
        function: _,
        declare: _,
    } = fn_decl;

    match hoist {
        Hoist::Everything => {
            let fn_ref = scope.declare_mutable(sym.clone());
            let init_ref = ir::Ref::new("_ini");
            vec![
                ir::Stmt::Expr {
                    target: init_ref.clone(),
                    expr: ir::Expr::Undefined,
                },
                ir::Stmt::DeclareMutable {
                    target: fn_ref,
                    val: init_ref,
                },
            ]
        }
        Hoist::OnlyLetConst => vec![],
    }
}

fn hoist_variable_declaration(
    var_decl: &ast::VarDecl,
    scope: &mut scope::Ast,
    hoist: &Hoist,
    level: &Level,
) -> Vec<ir::Stmt> {
    let ast::VarDecl {
        kind,
        decls,
        span: _,
        declare: _,
    } = var_decl;
    match (hoist, level, kind) {
        (Hoist::Everything, _, ast::VarDeclKind::Var) => decls
            .iter()
            .flat_map(
                |ast::VarDeclarator {
                     name,
                     init: _,
                     span: _,
                     definite: _,
                 }| {
                    let name = pat_to_ident(name);
                    match scope.get_mutable_in_current(name) {
                        None => {
                            let var_ref = scope.declare_mutable(name.clone());
                            let init_ref = ir::Ref::new("_ini");
                            vec![
                                ir::Stmt::Expr {
                                    target: init_ref.clone(),
                                    expr: ir::Expr::Undefined,
                                },
                                ir::Stmt::DeclareMutable {
                                    target: var_ref,
                                    val: init_ref,
                                },
                            ]
                        }
                        Some(_) => vec![],
                    }
                },
            )
            .collect(),
        (_, Level::First, ast::VarDeclKind::Let) | (_, Level::First, ast::VarDeclKind::Const) => {
            decls.iter().for_each(
                |ast::VarDeclarator {
                     name,
                     init: _,
                     span: _,
                     definite: _,
                 }| {
                    let name = pat_to_ident(name);
                    // declare to reserve name, in case some function above the declaration uses it
                    scope.declare_mutable(name.clone());
                },
            );
            vec![]
        }
        (Hoist::OnlyLetConst, _, ast::VarDeclKind::Var)
        | (_, Level::Below, ast::VarDeclKind::Let)
        | (_, Level::Below, ast::VarDeclKind::Const) => vec![],
    }
}

fn pat_to_ident(pat: &ast::Pat) -> &JsWord {
    match pat {
        ast::Pat::Ident(ast::Ident {
            sym,
            span: _,
            type_ann: _,
            optional: _,
        }) => sym,
        ast::Pat::Array(_)
        | ast::Pat::Object(_)
        | ast::Pat::Rest(_)
        | ast::Pat::Assign(_)
        | ast::Pat::Expr(_) => unimplemented!("complex patterns not supported"),
        ast::Pat::Invalid(_) => unreachable!(),
    }
}
