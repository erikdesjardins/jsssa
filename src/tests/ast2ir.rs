use crate::ast2ir;
use crate::ir;
use crate::parse;
use crate::swc_globals;

macro_rules! case {
    ( $name:ident, $string:expr ) => {
        #[test]
        fn $name() {
            swc_globals::with(|g| {
                let (ast, _) = parse::parse(g, $string).unwrap();
                let ir = ast2ir::convert(g, ast);
                let ppr = ir::print(g, &ir);
                insta::assert_snapshot_matches!(stringify!($name), ppr, $string);
            });
        }
    };
}

case!(
    basic,
    r#"
    function f(x) {
        while (true);
        x = y.bar;
        z.foo = x ? true : 'hi';
        return +[1 || x, { x }, f + 1, ++g];
    }
    f(1), true;
"#
);

case!(
    object_props,
    r#"
    o.x;
    o[x];
    o.x = 1;
    o[x] = 1;
    delete o.x;
    delete o[x];
"#
);

case!(
    var_reassignment_1,
    r#"
    var x = 1;
    const y = 1;
    {
        var x = 2;
        const y = 2;
    }
"#
);

case!(
    var_reassignment_2,
    r#"
    var x = 1;
    const y = 1;
    (function() {
        var x = 2;
        const y = 2;
    });
"#
);

case!(
    var_scoping_1,
    r#"
    var x = 1;
    {
        var x = 2;
    }
    x = 3;
    var x = 4;
"#
);

case!(
    var_scoping_2,
    r#"
    var x = 1;
    {
        let x = 2;
        x = 3;
    }
    x = 4;
"#
);

case!(
    var_hoisting_1,
    r#"
    x = 1;
    {
        var x = 2;
    }
    x = 3;
"#
);

case!(
    var_hoisting_2,
    r#"
    x = 1;
    {
        for (var x of 2);
    }
    x = 3;
"#
);
