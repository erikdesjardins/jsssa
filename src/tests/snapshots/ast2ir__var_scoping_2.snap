---
created: "2019-03-29T07:15:11.879629600Z"
creator: insta@0.7.1
source: src/tests/ast2ir.rs
expression: "\n    var x = 1;\n    {\n        let x = 2;\n        x = 3;\n    }\n    x = 4;\n"
---
_ini = <void>
x :- _ini
_ini$1 = 1
x <- _ini$1
<block>:
    _ini$2 = 2
    x$1 :- _ini$2
    _val = 3
    x$1 <- _val
    <dead> = _val
_val = 4
x <- _val
<dead> = _val
